pageserver/src/tenant/timeline.rs

```diff
@@ -434,6 +434,49 @@ pub struct Timeline {
 
     /// Cf. [`crate::tenant::CreateTimelineIdempotency`].
     pub(crate) create_idempotency: crate::tenant::CreateTimelineIdempotency,
+
+    pub(crate) page_trace: ArcSwap<Option<PageTrace>>,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub(crate) struct PageTraceEvent {
+    pub(crate) key: CompactKey,
+    pub(crate) effective_lsn: Lsn,
+    pub(crate) time: SystemTime,
+}
+
+/// When one of these is instantiated for a tenant, it will be used to record fine-grained
+/// history of getpage@lsn requests.
+pub(crate) struct PageTrace {
+    size_limit: u64,
+    size: AtomicU64,
+    trace_tx: tokio::sync::mpsc::UnboundedSender<PageTraceEvent>,
+}
+
+impl PageTrace {
+    pub(crate) fn new(
+        size_limit: u64,
+    ) -> (Self, tokio::sync::mpsc::UnboundedReceiver<PageTraceEvent>) {
+        let (trace_tx, trace_rx) = tokio::sync::mpsc::unbounded_channel();
```

Comment:
nit: we could also use a buffered channel with the max size here, to avoid the size accounting.